packages/ra-core/src/form/useWarnWhenUnsavedChanges.tsx

```diff
@@ -14,66 +13,82 @@ export const useWarnWhenUnsavedChanges = (
     formRootPathname?: string,
     control?: Control
 ) => {
-    // react-router v6 does not yet provide a way to block navigation
-    // This is planned for a future release
-    // See https://github.com/remix-run/react-router/issues/8139
-    const navigator = useContext(UNSAFE_NavigationContext).navigator as History;
-    const location = useLocation();
     const translate = useTranslate();
-    const { isSubmitSuccessful, isSubmitting, dirtyFields } = useFormState(
+    const { isSubmitSuccessful, dirtyFields } = useFormState(
         control ? { control } : undefined
     );
     const isDirty = Object.keys(dirtyFields).length > 0;
-    const initialLocation = useRef(formRootPathname || location.pathname);
+    const [shouldNotify, setShouldNotify] = useState(false);
+
+    const shouldNotBlock = !enable || !isDirty || isSubmitSuccessful;
+
+    const blocker = useBlocker(({ currentLocation, nextLocation }) => {
+        if (shouldNotBlock) return false;
+
+        // Also check if the new location is inside the form
+        const initialLocation = formRootPathname || currentLocation.pathname;
+        const newLocationIsInsideCurrentLocation = nextLocation.pathname.startsWith(
+            initialLocation
+        );
+        const newLocationIsShowView = nextLocation.pathname.startsWith(
+            `${initialLocation}/show`
+        );
+        const newLocationIsInsideForm =
+            newLocationIsInsideCurrentLocation && !newLocationIsShowView;
+        if (newLocationIsInsideForm) return false;
+
+        return true;
+    });
 
     useEffect(() => {
-        if (!enable || !isDirty) return;
-        if (!navigator.block) {
-            if (process.env.NODE_ENV !== 'production') {
-                console.warn(
-                    'warnWhenUnsavedChanged is not compatible with react-router >= 6.4. If you need this feature, please downgrade react-router to 6.3.0'
-                );
+        if (blocker.state === 'blocked') {
+            // Corner case: the blocker might be triggered by a redirect in the onSuccess side effect,
+            // happening during the same tick the form is reset after a successful save.
+            // In that case, the blocker will block but shouldNotBlock will be true one tick after.
+            // If we are in that case, we can proceed immediately.
+            if (shouldNotBlock) {
+                blocker.proceed();
+                return;
             }
-            return;
+
+            setShouldNotify(true);
         }
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [blocker.state]);
```

Comment:
shouldNotBlock is missing here. It seems you know why but the reader of that code won't. Would you mind explaining it in a comment?